Add upgrade guide for Middleware. (#37382)
vercel/next.js: errors/middleware-upgrade-guide.md

# Middleware Upgrade Guide

> As we work on improving Middleware for General Availability (GA), we've made some changes to the Middleware APIs (and how you define Middleware in your application) based on your feedback.

Reviewer: I don't think the parentheses are necessary. Makes it a better easier to read without?
```suggestion
As we work on improving Middleware for General Availability (GA), we've made some changes to the Middleware APIs and to how you define Middleware in your application, based on your feedback.
```

Author: "to how you" feels strange - I'm open to dropping the parenthesis but might need a different revision.

This upgrade guide will help you understand the changes and how to migrate your existing Middleware to the new API. This guide is for Next.js customers who:

- Currently use the beta Next.js Middleware features
- Choose to upgrade to the next stable version of Next.js

> > **Note:** For customers using Next.js on Vercel, your existing deploys using Middlware will continue to work. Existing sites can be redeployed without needing to upgrade. Before upgrading to the next stable version of Next.js, you will need to follow the upgrade guide below.

Reviewer: Is there value in being explicit about adding the version of Next.js here so people can be super sure if they have to take action and when? 

Also, typo and text suggestion

```suggestion
> **Note:** For customers using Next.js on Vercel, your existing deploys using Middleware will continue to work. Existing sites can be redeployed without needing to upgrade. Before upgrading to the next stable version of Next.js, you will need to follow this upgrade guide in order for your existing deploy to continue to work.
```

Author: Replacing with https://github.com/vercel/next.js/pull/37382/files#r888000812

> > **Note:** For customers using Next.js on Vercel, your existing deploys using Middlware will continue to work. Existing sites can be redeployed without needing to upgrade. Before upgrading to the next stable version of Next.js, you will need to follow the upgrade guide below.

Reviewer 2: I agree that we should be explicit about the version.

I wonder if we should pull this out into its own section instead of a note? As a developer, I'd want to understand what the immediate impact of these changes are to me first, and then understand how I adapt to these changes.

More text suggestions. 🙃 

```suggestion
## Using Next.js Middleware on Vercel

If you're using Next.js on Vercel, your existing deploys using Middleware will continue to work, and you can continue to deploy your site using Middleware. When you upgrade your site to the next stable version of Next.js, you will need to follow this upgrade guide to update your Middleware.
```

## Breaking changes

1. [No Nested Middleware](#no-nested-middleware)
2. [No Response Body](#no-response-body)
3. [Cookies API Revamped](#cookies-api-revamped)
4. [No More Page Match Data](#no-more-page-match-data)
5. [Executing Middleware on Internal Next.js Requests](#executing-middleware-on-internal-next.js-requests)

## No Nested Middleware

### Summary of changes

- Define a single Middleware file at the root of your project
- No need to prefix the file with an underscore
- A custom matcher can be used to define matching routes using an exported config object

### Explanation

Previously, you could create a `_middleware.js` file under the `pages` directory at any level. Middleware execution was based on the file path where it was created. Beta customers found this route matching confusing. For example:

- Middleware in `pages/dashboard/_middleware.ts`

> - Middlware in `pages/dashboard/users/_middleware.ts`

Reviewer 3: ```suggestion
- Middleware in `pages/dashboard/users/_middleware.ts`
```

- A request to `/dashboard/users/*` **would match both.**

Based on customer feedback, we have replaced this API with a single root Middleware.

### How to upgrade

You should declare **one single Middleware file** in your application, which should be located at the root of the project directory (**not** inside of the `pages` directory) and named without an `_` prefix. Your Middleware file can still have either a `.ts` or `.js` extension.

Middleware will be invoked for **every route in the app**, and a custom matcher can be used to define matching filters. The following is an example for a Middleware that triggers for `/about/*`, the custom matcher is defined in an exported config object:

```typescript
// middleware.ts
import type { NextRequest } from 'next/server'

export function middleware(request: NextRequest) {
  return NextResponse.rewrite(new URL('/about-2', request.url))
}
// config with custom matcher
export const config = {

>   pathname: '/about/:path*',

Reviewer 4: Decided name is `matcher`
```suggestion
  matcher: '/about/:path*',
```

}
```

> You can also use a conditional statement to only run the Middleware when it matches the `about/*` path:

Reviewer 3: I guess the difference between the two approaches is that the middleware in the approach below gets invoked much more often (for every request, vs only for requests to `/about/*`)?

Might be nice to call that difference out explicitly

Author: Updated

> You can also use a conditional statement to only run the Middleware when it matches the `about/*` path:

Reviewer 4: ```suggestion
You can also use a conditional statement to only run the Middleware when it matches either the `about/*` or `dashboard/*` paths. The following example shows how you can merge two previously nested Middleware:
```

```typescript
// <root>/middleware.js
import type { NextRequest } from 'next/server'

export function middleware(request: NextRequest) {
  if (request.nextUrl.pathname.startsWith('/about')) {
    // This logic is only applied to /about
  }

  if (request.nextUrl.pathname.startsWith('/dashboard')) {
    // This logic is only applied to /dashboard
  }
}
```

## No Response Body

### Summary of changes

- Middleware can no longer respond with a body
- If your Middleware _does_ respond with a body, a runtime error will be thrown
- Migrate to using `rewrites`/`redirects` to pages that handle authorization

### Explanation

> Beta customers had explored using Middleware to handle authorization for their application. However, to ensure both the HTML and data payload (JSON file) are protected, we commend checking authorization at the Page level.

Reviewer 4: ```suggestion
Beta customers had explored using Middleware to handle authorization for their application. However, to ensure both the HTML and data payload (JSON file) are protected, we recommend checking authorization at the Page level.
```

> Beta customers had explored using Middleware to handle authorization for their application. However, to ensure both the HTML and data payload (JSON file) are protected, we commend checking authorization at the Page level.
> 
> To help ensure security, we are removing the ability to send response bodies in Middleware. This ensures that Middleware is only used to `rewrite`, `redirect`, or modify the incoming request (e.g. [setting cookies](#cookies-api-revamped)).

Reviewer: It feels like these paragraphs should be swapped. The second paragraph here shows the "what" and then first the "why". I think it might be more logically to give the "what" up front.

```suggestion
To help ensure security, **we are removing the ability to send response bodies in Middleware**. This ensures that Middleware is only used to `rewrite`, `redirect`, or modify the incoming request (e.g. [setting cookies](#cookies-api-revamped)).

Beta customers had explored using Middleware to handle authorization for their application. However, to ensure both the HTML and data payload (JSON file) are protected, we recommend checking authorization at the page level.
```

The following patterns will no longer work:

```js
new Response('a text value')
new Response(streamOrBuffer)

> new Response(JSON.stringify(obj), { headers: 'application/json' })

Reviewer 3: Should `NextResponse.json()` be called out as well? Not sure

Reviewer 5: Yeah, it is now an error to use it

```

### How to upgrade

> For cases where Middleware is used to respond (such as authorization), you should migrate to use `rewrites`/`redirects` to Pages that show an authorization error, login forms, or to an API Route.

Reviewer: Should "Pages" be capitalized?

> For cases where Middleware is used to respond (such as authorization), you should migrate to use `rewrites`/`redirects` to Pages that show an authorization error, login forms, or to an API Route.

Author: Yes - it's a Next.js product

For cases where Middleware is used to respond (such as authorization), you should migrate to use `rewrites`/`redirects` to Pages that show an authorization error, login forms, or to an API Route.

> 

Reviewer 4: ```suggestion
#### Before
	
	```typescript
	// middleware.ts
	import { NextResponse } from 'next/server'
	import type { NextRequest } from 'next/server'
	import { isAuthValid } from './lib/auth'
	
	export function middleware(request: NextRequest) {
	// Example function to validate auth
	  if (isAuthValid(req)) {
	    return NextResponse.next()
	  }
	  return NextResponse.json({ message: 'Auth required' }, { status: 401 })
	}
	
#### After
```

> ```typescript
> // middleware.ts
> import { NextResponse } from 'next/server'
> import type { NextRequest } from 'next/server'
> import { credentialsValid } from './lib/auth'
> 
> export function middleware(request: NextRequest) {
>   const basicAuth = request.headers.get('authorization')
> 
>   if (basicAuth) {
>     const auth = basicAuth.split(' ')[1]
>     const [user, pwd] = atob(auth).split(':')
>     // Example function to validate auth
>     if (credentialsValid(user, pwd)) {
>       return NextResponse.next()
>     }
>   }
> 
>   const loginUrl = new URL('/login', request.url)
>   loginUrl.searchParams.set('from', request.nextUrl.pathname)
> 
>   return NextResponse.redirect(loginUrl)
> }
> ```

Reviewer 4: ```suggestion

#### Before

```typescript
// middleware.ts
import { NextResponse } from 'next/server'
import type { NextRequest } from 'next/server'
import { isAuthValid } from './lib/auth'

export function middleware(request: NextRequest) {
// Example function to validate auth
  if (isAuthValid(req)) {
    return NextResponse.next()
  }
  return NextResponse.json({ message: 'Auth required' }, { status: 401 })
}
```

#### After

```typescript
// middleware.ts
import { NextResponse } from 'next/server'
import type { NextRequest } from 'next/server'
import { credentialsValid } from './lib/auth'

export function middleware(request: NextRequest) {
  const basicAuth = request.headers.get('authorization')

  if (basicAuth) {
    const auth = basicAuth.split(' ')[1]
    const [user, pwd] = atob(auth).split(':')
    // Example function to validate auth
    if (credentialsValid(user, pwd)) {
      return NextResponse.next()
    }
  }

  const loginUrl = new URL('/login', request.url)
  loginUrl.searchParams.set('from', request.nextUrl.pathname)

  return NextResponse.redirect(loginUrl)
}
```
```

## Cookies API Revamped

### Summary of changes

**Added**

- `cookie.delete`
- `cookie.set`

> - `cookie.getAll`

Reviewer 4: ```suggestion
- `cookie.getWithOptions`
```

Reviewer 6: I'm not sure how to write this, but since the internal [`Cookies`](https://github.com/vercel/next.js/blob/canary/packages/next/server/web/spec-extension/cookies.ts#L43=) implementation extends from Map, it inherits the rest of methods, such as [entries](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Map/entries) or [values](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Map/entries), etc

Author: I'll add this

**Removed:**

- `cookie`
- `cookies`
- `clearCookie`

### Explanation

Based on beta feedback, we are changing the Cookies API in `NextRequest` and `NextResponse` to align more to a `get`/`set` model.

### How to upgrade

`NextResponse` now has a `cookies` instance with:

- `cookie.delete`
- `cookie.set`

> - `cookie.getAll`

Reviewer 4: ```suggestion
- `cookie.getWithOptions`
```

    const { search, protocol, host } = request.nextUrl
    const url = new URL(`${protocol}//${locale}.${host}/${slug}${search}`)
    return NextResponse.redirect(url)
  }
}
```

#### After

```typescript
// middleware.ts
import { NextRequest, NextResponse } from 'next/server'

const PATTERNS = [
  [
    new URLPattern({ pathname: '/:locale/:slug' }),
    ({ pathname }) => pathname.groups,
  ],
]

const params = (url) => {
  const input = url.split('?')[0]
  let result = {}

  for (const [pattern, handler] of PATTERNS) {
    const patternResult = pattern.exec(input)
    if (patternResult !== null && 'pathname' in patternResult) {
      result = handler(patternResult)
      break
    }
  }
  return result
}

export function middleware(request: NextRequest) {
  const { locale, slug } = params(request.url)

  if (locale && slug) {
    const { search, protocol, host } = request.nextUrl
    const url = new URL(`${protocol}//${locale}.${host}/${slug}${search}`)
    return NextResponse.redirect(url)
  }
}
```

## Executing Middleware on Internal Next.js Requests

### Summary of changes

- Middleware will be executed for _all_ requests, including `_next`

### Explanation

> Currently, we do not execute Middleware for `_next` requests, due the authorization use cases.

Reviewer 3: I had to re-read that section twice. I think changing this to "So far, we did not..." could make it clearer?

  if (request.nextUrl.pathname.startsWith('/about')) {
    // This logic is only applied to /about
  }

  if (request.nextUrl.pathname.startsWith('/dashboard')) {
    // This logic is only applied to /dashboard
  }
}
```

## No Response Body

### Summary of changes

- Middleware can no longer respond with a body
- If your Middleware _does_ respond with a body, a runtime error will be thrown
- Migrate to using `rewrites`/`redirects` to pages that handle authorization

### Explanation

Beta customers had explored using Middleware to handle authorization for their application. However, to ensure both the HTML and data payload (JSON file) are protected, we recommend checking authorization at the Page level.

To help ensure security, we are removing the ability to send response bodies in Middleware. This ensures that Middleware is only used to `rewrite`, `redirect`, or modify the incoming request (e.g. [setting cookies](#cookies-api-revamped)).

The following patterns will no longer work:

```js
new Response('a text value')
new Response(streamOrBuffer)
new Response(JSON.stringify(obj), { headers: 'application/json' })
```

### How to upgrade

> For cases where Middleware is used to respond (such as authorization), you should migrate to use `rewrites`/`redirects` to Pages that show an authorization error, login forms, or to an API Route.

Reviewer: The user action here is similar to https://github.com/vercel/next.js/pull/37382/files#diff-df4a93b5f2399985c81d6fb43adc2815bd39f190a0f770724bd7caf6a3628a73R292 for authoriztion. Is there some way to link them both together a bit more clearly?

Author: Done - added a link back at the bottom to this section

    sameSite: 'strict',
    domain: 'example.com',
  })
  // clear the `cookie`
  response.clearCookie('hello')

  return response
}
```

#### After

```typescript
// middleware.ts
export function middleware() {
  const response = new NextResponse()

  // set a cookie
  response.cookies.set('vercel', 'fast')

  // set another cookie with options
  response.cookies.set('nextjs', 'awesome', { path: '/test' })

  // get all the details of a cookie
  const [value, options] = response.cookies.getWithOptions('vercel')
  console.log(value) // => 'fast'
  console.log(options) // => { Path: '/test' }

  // delete a cookie means mark it as expired
  response.cookies.delete('vercel')

  // clear all cookies means mark all of them as expired
  response.cookies.clear()
}
```

## No More Page Match Data

### Summary of changes

- Use [`URLPattern`](https://developer.mozilla.org/en-US/docs/Web/API/URLSearchParams) to check if a Middleware is being invoked for a certain page match

### Explanation

Currently, Middleware estimates whether you are serving an asset of a Page based on the Next.js routes manifest (internal configuration). This value is surfaced through `request.page`.

> To make this more accurate, we are recommending to use the web standard `URLPattern` API.

Reviewer: Ambiguous "this"

- A custom matcher can be used to define matching routes using an exported config object

### Explanation

Previously, you could create a `_middleware.ts` file under the `pages` directory at any level. Middleware execution was based on the file path where it was created. Beta customers found this route matching confusing. For example:

- Middleware in `pages/dashboard/_middleware.ts`
- Middleware in `pages/dashboard/users/_middleware.ts`
- A request to `/dashboard/users/*` **would match both.**

Based on customer feedback, we have replaced this API with a single root Middleware.

### How to upgrade

You should declare **one single Middleware file** in your application, which should be located at the root of the project directory (**not** inside of the `pages` directory), and named **without** an `_` prefix. Your Middleware file can still have either a `.ts` or `.js` extension.

Middleware will be invoked for **every route in the app**, and a custom matcher can be used to define matching filters. The following is an example for a Middleware that triggers for `/about/*`, the custom matcher is defined in an exported config object:

```typescript
// middleware.ts

> import type { NextRequest, NextResponse } from 'next/server'

Reviewer 7: I think the "type" on import here will break referencing this on line 51?

Reviewer 4: Yeah they should be separated, good catch

> import type { NextRequest, NextResponse } from 'next/server'

Reviewer 8: ```suggestion
import { NextResponse } from 'next/server'
import type { NextRequest } from 'next/server'
```

export function middleware(request: NextRequest) {
  return NextResponse.rewrite(new URL('/about-2', request.url))
}
// Config with custom matcher
// Pending: https://github.com/vercel/next.js/pull/37177
export const config = {

>   matcher: '/about/:path*',

Reviewer 5: Arrays are now supported

}
```

> 

Reviewer 4: #### Match paths based on custom matcher config
To decide which route the middleware should be run on, you can use a custom matcher config to filter on specific paths. The matcher property can be used to define either a single path, or using an array syntax, multiple paths.

##### Match a single path

```js
export const config = {
  matcher: '/about/:path*'
}
```

##### Match multiple paths

```js
export const config = {
  matcher: ['/about/:path*', '/dashboard/:path*']
}
```

- `cookie.getWithOptions`

As well as other extended methods from `Map`.

#### Before

```javascript
import { NextResponse } from 'next/server'
import type { NextRequest } from 'next/server'

export function middleware(request: NextRequest) {
  // create an instance of the class to access the public methods. This uses `next()`,
  // you could use `redirect()` or `rewrite()` as well
  let response = NextResponse.next()
  // get the cookies from the request
  let cookieFromRequest = request.cookies['my-cookie']
  // set the `cookie`
  response.cookie('hello', 'world')
  // set the `cookie` with options
  const cookieWithOptions = response.cookie('hello', 'world', {
    path: '/',
    maxAge: 1000 * 60 * 60 * 24 * 7,
    httpOnly: true,
    sameSite: 'strict',
    domain: 'example.com',
  })
  // clear the `cookie`
  response.clearCookie('hello')

  return response
}
```

#### After

```typescript
// middleware.ts
export function middleware() {
  const response = new NextResponse()

  // set a cookie
  response.cookies.set('vercel', 'fast')

  // set another cookie with options
  response.cookies.set('nextjs', 'awesome', { path: '/test' })

  // get all the details of a cookie

>   const [value, options] = response.cookies.getWithOptions('vercel')

Reviewer 4: Tested this and it doesnt return a destructured array, rather an object
```suggestion
  const {value, options} = response.cookies.getWithOptions('vercel')
```

Reviewer 4: cc @Kikobeats

Reviewer 6: That's right @molebox, thanks for fixing 🙂

  console.log(value) // => 'fast'
  console.log(options) // => { Path: '/test' }

  // delete a cookie means mark it as expired
  response.cookies.delete('vercel')

  // clear all cookies means mark all of them as expired
  response.cookies.clear()
}
```

## No More Page Match Data

### Summary of changes

> - Use [`URLPattern`](https://developer.mozilla.org/en-US/docs/Web/API/URLSearchParams) to check if a Middleware is being invoked for a certain page match

Reviewer 9: ```suggestion
- Use [`URLPattern`](https://developer.mozilla.org/en-US/docs/Web/API/URLPattern) to check if a Middleware is being invoked for a certain page match
```

5. [Executing Middleware on Internal Next.js Requests](#executing-middleware-on-internal-nextjs-requests)

## No Nested Middleware

### Summary of changes

- Define a single Middleware file at the root of your project
- No need to prefix the file with an underscore
- A custom matcher can be used to define matching routes using an exported config object

### Explanation

Previously, you could create a `_middleware.ts` file under the `pages` directory at any level. Middleware execution was based on the file path where it was created. Beta customers found this route matching confusing. For example:

- Middleware in `pages/dashboard/_middleware.ts`
- Middleware in `pages/dashboard/users/_middleware.ts`
- A request to `/dashboard/users/*` **would match both.**

Based on customer feedback, we have replaced this API with a single root Middleware.

### How to upgrade

You should declare **one single Middleware file** in your application, which should be located at the root of the project directory (**not** inside of the `pages` directory), and named **without** an `_` prefix. Your Middleware file can still have either a `.ts` or `.js` extension.

> Middleware will be invoked for **every route in the app**, and a custom matcher can be used to define matching filters. The following is an example for a Middleware that triggers for `/about/*`, the custom matcher is defined in an exported config object:

Reviewer 10: Below the request will be executing for both `/about/*` and `/dashboard/:path*` but the line only mentions `/about/*` :thonk:

> Middleware will be invoked for **every route in the app**, and a custom matcher can be used to define matching filters. The following is an example for a Middleware that triggers for `/about/*`, the custom matcher is defined in an exported config object:

Author: ```suggestion
Middleware will be invoked for **every route in the app**, and a custom matcher can be used to define matching filters. The following is an example for a Middleware that triggers for `/about/*` and `/dashboard/:path*`, the custom matcher is defined in an exported config object:
```

  matcher: ['/about/:path*', '/dashboard/:path*'],
}
```

While the config option is preferred since it doesn't get invoked on every request, you can also use conditional statements to only run the Middleware when it matches specific paths. The following example shows how you can merge two previously nested Middleware:

```typescript
// <root>/middleware.js
import type { NextRequest } from 'next/server'

export function middleware(request: NextRequest) {
  if (request.nextUrl.pathname.startsWith('/about')) {
    // This logic is only applied to /about
  }

  if (request.nextUrl.pathname.startsWith('/dashboard')) {
    // This logic is only applied to /dashboard
  }
}
```

## No Response Body

### Summary of changes

- Middleware can no longer respond with a body
- If your Middleware _does_ respond with a body, a runtime error will be thrown
- Migrate to using `rewrites`/`redirects` to pages/APIs handling a response

### Explanation

To help ensure security, we are removing the ability to send response bodies in Middleware. This ensures that Middleware is only used to `rewrite`, `redirect`, or modify the incoming request (e.g. [setting cookies](#cookies-api-revamped)).

> Beta customers had explored using Middleware to handle authorization for their application. However, to ensure both the HTML and data payload (JSON file) are protected, we recommend checking authorization at the page level.

Reviewer 10: You can still handle it right? Like checking the token from cookies and rewrite / redirect to a page if the token is not valid

> 
> Beta customers had explored using Middleware to handle authorization for their application. However, to ensure both the HTML and data payload (JSON file) are protected, we recommend checking authorization at the page level.

Author: ```suggestion

```

# Middleware Upgrade Guide

As we work on improving Middleware for General Availability (GA), we've made some changes to the Middleware APIs (and how you define Middleware in your application) based on your feedback.

> This upgrade guide will help you understand the changes and how to migrate your existing Middleware to the new API. The guide is for Next.js customers who:

Author: ```suggestion
This upgrade guide will help you understand the changes and how to migrate your existing Middleware to the new API. The guide is for Next.js developers who:
```

- Currently use the beta Next.js Middleware features
- Choose to upgrade to the next stable version of Next.js

## Using Next.js Middleware on Vercel

If you're using Next.js on Vercel, your existing deploys using Middleware will continue to work, and you can continue to deploy your site using Middleware. When you upgrade your site to the next stable version of Next.js (`v12.2`), you will need to follow this upgrade guide to update your Middleware.

## Breaking changes

1. [No Nested Middleware](#no-nested-middleware)
2. [No Response Body](#no-response-body)
3. [Cookies API Revamped](#cookies-api-revamped)
4. [No More Page Match Data](#no-more-page-match-data)

> 5. [Executing Middleware on Internal Next.js Requests](#executing-middleware-on-internal-next.js-requests)

Reviewer 4: this link doesnt work

> 5. [Executing Middleware on Internal Next.js Requests](#executing-middleware-on-internal-next.js-requests)

Author: ```suggestion
5. [Executing Middleware on Internal Next.js Requests](#executing-middleware-on-internal-nextjs-requests)
```

## No Nested Middleware

### Summary of changes

- Define a single Middleware file at the root of your project
- No need to prefix the file with an underscore
- A custom matcher can be used to define matching routes using an exported config object

### Explanation

Previously, you could create a `_middleware.js` file under the `pages` directory at any level. Middleware execution was based on the file path where it was created. Beta customers found this route matching confusing. For example:

- Middleware in `pages/dashboard/_middleware.ts`
- Middleware in `pages/dashboard/users/_middleware.ts`
- A request to `/dashboard/users/*` **would match both.**

Based on customer feedback, we have replaced this API with a single root Middleware.

### How to upgrade

> You should declare **one single Middleware file** in your application, which should be located at the root of the project directory (**not** inside of the `pages` directory), and named without an `_` prefix. Your Middleware file can still have either a `.ts` or `.js` extension.

Reviewer 4: Lets hammer it home
```suggestion
You should declare **one single Middleware file** in your application, which should be located at the root of the project directory (**not** inside of the `pages` directory), and named **without** an `_` prefix. Your Middleware file can still have either a `.ts` or `.js` extension.
```

Middleware will be invoked for **every route in the app**, and a custom matcher can be used to define matching filters. The following is an example for a Middleware that triggers for `/about/*`, the custom matcher is defined in an exported config object:

```typescript
// middleware.ts
import type { NextRequest } from 'next/server'

export function middleware(request: NextRequest) {
  return NextResponse.rewrite(new URL('/about-2', request.url))
}

> // config with custom matcher

Author: ```suggestion
// Config with custom matcher
// Pending: https://github.com/vercel/next.js/pull/37177
```

> export const config = {

Author: Thank you!

  matcher: '/about/:path*',
}
```

While the config option is preferred since it doesn't get invoked on every request, you can also use conditional statements to only run the Middleware when it matches specific paths. The following example shows how you can merge two previously nested Middleware:

```typescript
// <root>/middleware.js
import type { NextRequest } from 'next/server'

export function middleware(request: NextRequest) {
  if (request.nextUrl.pathname.startsWith('/about')) {
    // This logic is only applied to /about
  }

  if (request.nextUrl.pathname.startsWith('/dashboard')) {
    // This logic is only applied to /dashboard
  }
}
```

## No Response Body

### Summary of changes

- Middleware can no longer respond with a body
- If your Middleware _does_ respond with a body, a runtime error will be thrown
- Migrate to using `rewrites`/`redirects` to pages that handle authorization

### Explanation

To help ensure security, we are removing the ability to send response bodies in Middleware. This ensures that Middleware is only used to `rewrite`, `redirect`, or modify the incoming request (e.g. [setting cookies](#cookies-api-revamped)).

> Beta customers had explored using Middleware to handle authorization for their application. However, to ensure both the HTML and data payload (JSON file) are protected, we recommend checking authorization at the Page level.

Reviewer 6: ```suggestion
Beta customers had explored using Middleware to handle authorization for their application. However, to ensure both the HTML and data payload (JSON file) are protected, we recommend checking authorization at the page level.
```

The following patterns will no longer work:

```js
new Response('a text value')
new Response(streamOrBuffer)
new Response(JSON.stringify(obj), { headers: 'application/json' })
```

### How to upgrade

> For cases where Middleware is used to respond (such as authorization), you should migrate to use `rewrites`/`redirects` to Pages that show an authorization error, login forms, or to an API Route.

Reviewer 6: ```suggestion
For cases where Middleware is used to respond (such as authorization), you should migrate to use `rewrites`/`redirects` to pages that show an authorization error, login forms, or to an API Route.
```

| Added                   | Removed       |
| ----------------------- | ------------- |
| `cookie.set`            | `cookie`      |
| `cookie.delete`         | `clearCookie` |
| `cookie.getWithOptions` | `cookies`     |

### Explanation

Based on beta feedback, we are changing the Cookies API in `NextRequest` and `NextResponse` to align more to a `get`/`set` model. The `Cookies` API extends Map, including methods like [entries](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Map/entries) and [values](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Map/entries).

### How to upgrade

`NextResponse` now has a `cookies` instance with:

- `cookie.delete`
- `cookie.set`
- `cookie.getWithOptions`

As well as other extended methods from `Map`.

#### Before

```javascript
import { NextResponse } from 'next/server'
import type { NextRequest } from 'next/server'

export function middleware(request: NextRequest) {
  // create an instance of the class to access the public methods. This uses `next()`,
  // you could use `redirect()` or `rewrite()` as well
  let response = NextResponse.next()
  // get the cookies from the request

>   let cookieFromRequest = request.cookies['my-cookie']

Author: ```suggestion
  let cookieFromRequest = request.cookies.get('my-cookie')
```

Reviewer 8: From a discussion with @Kikobeats: prior to the change, request and response cookies were both plain objects.
This snipper is in the "Before" paragraph, so it should **not** use `.get()` nor `.set()`

>   let cookieFromRequest = request.cookies['my-cookie']

Author: Good question 👍 Let me check.

>   let cookieFromRequest = request.cookies['my-cookie']

Author: Nope: https://github.com/vercel/next.js/pull/37382/files#r888992310

This upgrade guide will help you understand the changes and how to migrate your existing Middleware to the new API. The guide is for Next.js developers who:

- Currently use the beta Next.js Middleware features
- Choose to upgrade to the next stable version of Next.js

## Using Next.js Middleware on Vercel

If you're using Next.js on Vercel, your existing deploys using Middleware will continue to work, and you can continue to deploy your site using Middleware. When you upgrade your site to the next stable version of Next.js (`v12.2`), you will need to follow this upgrade guide to update your Middleware.

## Breaking changes

1. [No Nested Middleware](#no-nested-middleware)
2. [No Response Body](#no-response-body)
3. [Cookies API Revamped](#cookies-api-revamped)
4. [No More Page Match Data](#no-more-page-match-data)
5. [Executing Middleware on Internal Next.js Requests](#executing-middleware-on-internal-nextjs-requests)

## No Nested Middleware

### Summary of changes

- Define a single Middleware file at the root of your project
- No need to prefix the file with an underscore
- A custom matcher can be used to define matching routes using an exported config object

### Explanation

> Previously, you could create a `_middleware.js` file under the `pages` directory at any level. Middleware execution was based on the file path where it was created. Beta customers found this route matching confusing. For example:

Author: ```suggestion
Previously, you could create a `_middleware.ts` file under the `pages` directory at any level. Middleware execution was based on the file path where it was created. Beta customers found this route matching confusing. For example:
```

  // set another cookie with options
  response.cookies.set('nextjs', 'awesome', { path: '/test' })

  // get all the details of a cookie
  const [value, options] = response.cookies.getWithOptions('vercel')
  console.log(value) // => 'fast'
  console.log(options) // => { Path: '/test' }

  // delete a cookie means mark it as expired
  response.cookies.delete('vercel')

  // clear all cookies means mark all of them as expired
  response.cookies.clear()
}
```

## No More Page Match Data

### Summary of changes

- Use [`URLPattern`](https://developer.mozilla.org/en-US/docs/Web/API/URLSearchParams) to check if a Middleware is being invoked for a certain page match

### Explanation

Currently, Middleware estimates whether you are serving an asset of a Page based on the Next.js routes manifest (internal configuration). This value is surfaced through `request.page`.

To make page and asset matching more accurate, we are now using the web standard `URLPattern` API.

### How to upgrade

Use [`URLPattern`](https://developer.mozilla.org/en-US/docs/Web/API/URLPattern) to check if a Middleware is being invoked for a certain page match.

#### Before

```typescript
// middleware.ts

> import { NextRequest, NextResponse } from 'next/server'

Reviewer 4: NextRequest should be imported as a type, as a separate import. Although you can do it this way on latest TS version, for backwards compatibility we should show the "old" way and save ourselves having to explain and recommend updating TS versions

> import { NextRequest, NextResponse } from 'next/server'

Author: ```suggestion
import { NextResponse } from 'next/server'
import type { NextRequest } from 'next/server'
```

export function middleware(request: NextRequest) {
  const { params } = event.request.page
  const { locale, slug } = params

  if (locale && slug) {
    const { search, protocol, host } = request.nextUrl
    const url = new URL(`${protocol}//${locale}.${host}/${slug}${search}`)
    return NextResponse.redirect(url)
  }
}
```

#### After

```typescript
// middleware.ts

> import { NextRequest, NextResponse } from 'next/server'

Reviewer 4: See above about imports

> import { NextRequest, NextResponse } from 'next/server'

Author: ```suggestion
import { NextResponse } from 'next/server'
import type { NextRequest } from 'next/server'
```

### Explanation

Previously, you could create a `_middleware.ts` file under the `pages` directory at any level. Middleware execution was based on the file path where it was created. Beta customers found this route matching confusing. For example:

- Middleware in `pages/dashboard/_middleware.ts`
- Middleware in `pages/dashboard/users/_middleware.ts`
- A request to `/dashboard/users/*` **would match both.**

Based on customer feedback, we have replaced this API with a single root Middleware.

### How to upgrade

You should declare **one single Middleware file** in your application, which should be located at the root of the project directory (**not** inside of the `pages` directory), and named **without** an `_` prefix. Your Middleware file can still have either a `.ts` or `.js` extension.

Middleware will be invoked for **every route in the app**, and a custom matcher can be used to define matching filters. The following is an example for a Middleware that triggers for `/about/*`, the custom matcher is defined in an exported config object:

```typescript
// middleware.ts
import type { NextRequest } from 'next/server'

export function middleware(request: NextRequest) {

>   return NextResponse.rewrite(new URL('/about-2', request.url))

Reviewer 11: Is `NextResponse` in scope here?

}
// Config with custom matcher
// Pending: https://github.com/vercel/next.js/pull/37177

> export const config = {

Reviewer 11: Does/can `config` have a type?

Author: I will update this once the PR linked merges 😄

Beta customers had explored using Middleware to handle authorization for their application. However, to ensure both the HTML and data payload (JSON file) are protected, we recommend checking authorization at the page level.

The following patterns will no longer work:

```js
new Response('a text value')
new Response(streamOrBuffer)
new Response(JSON.stringify(obj), { headers: 'application/json' })
```

### How to upgrade

For cases where Middleware is used to respond (such as authorization), you should migrate to use `rewrites`/`redirects` to pages that show an authorization error, login forms, or to an API Route.

#### Before

```typescript
// pages/_middleware.ts
import { NextResponse } from 'next/server'
import type { NextRequest } from 'next/server'
import { isAuthValid } from './lib/auth'

export function middleware(request: NextRequest) {
  // Example function to validate auth
  if (isAuthValid(req)) {
    return NextResponse.next()
  }

  return NextResponse.json({ message: 'Auth required' }, { status: 401 })

> }

Reviewer 11: :chefskiss:

```

#### After

```typescript
// middleware.ts
import { NextResponse } from 'next/server'
import type { NextRequest } from 'next/server'

> import { credentialsValid } from './lib/auth'

Reviewer 8: ```suggestion
import { isAuthValid } from './lib/auth'
```

export function middleware(request: NextRequest) {

>   const basicAuth = request.headers.get('authorization')
> 
>   if (basicAuth) {
>     const auth = basicAuth.split(' ')[1]
>     const [user, pwd] = atob(auth).split(':')
>     // Example function to validate auth
>     if (credentialsValid(user, pwd)) {
>       return NextResponse.next()
>     }
>   }

Reviewer 8: This part is coming from an older example, and does not match the one in "Before" paragraph.

```suggestion
  // Example function to validate auth
  if (isAuthValid(req)) {
    return NextResponse.next()
  }
```

# Middleware Upgrade Guide

As we work on improving Middleware for General Availability (GA), we've made some changes to the Middleware APIs (and how you define Middleware in your application) based on your feedback.

This upgrade guide will help you understand the changes and how to migrate your existing Middleware to the new API. The guide is for Next.js developers who:

- Currently use the beta Next.js Middleware features

> - Choose to upgrade to the next stable version of Next.js

Author: ```suggestion
- Choose to upgrade to the next stable version of Next.js (`v12.2`)

You can start upgrading your Middleware usage today with the latest canary release (`npm i next@canary`).
```

  // set the `cookie`
  response.cookie('hello', 'world')
  // set the `cookie` with options
  const cookieWithOptions = response.cookie('hello', 'world', {
    path: '/',
    maxAge: 1000 * 60 * 60 * 24 * 7,
    httpOnly: true,
    sameSite: 'strict',
    domain: 'example.com',
  })
  // clear the `cookie`
  response.clearCookie('hello')

  return response
}
```

#### After

```typescript
// middleware.ts
export function middleware() {
  const response = new NextResponse()

  // set a cookie
  response.cookies.set('vercel', 'fast')

  // set another cookie with options
  response.cookies.set('nextjs', 'awesome', { path: '/test' })

  // get all the details of a cookie
  const {value, options} = response.cookies.getWithOptions('vercel')
  console.log(value) // => 'fast'
  console.log(options) // => { Path: '/test' }

  // deleting a cookie will mark it as expired
  response.cookies.delete('vercel')

  // clear all cookies means mark all of them as expired
  response.cookies.clear()
}
```
## New user agent helper

### Summary of changes

- User agent no longer available on the request object
- New helper introduced
- Reduces Middleware size by 17kB

### Explanation

> To help reduce the size of your Middleware, we have extracted the user agent from the request, and created a new helper to interact with the user agent object. 

Author: ```suggestion
To help reduce the size of your Middleware, we have extracted the user agent from the request object and created a new helper `userAgent`. 
```
